mlflow/store/artifact/s3_artifact_repo.py

```diff
@@ -97,26 +115,47 @@ def _get_s3_client(
         access_key_id=access_key_id,
         secret_access_key=secret_access_key,
         session_token=session_token,
+        region_name=region_name,
     )
 
 
-class S3ArtifactRepository(ArtifactRepository):
+class S3ArtifactRepository(CloudArtifactRepository):
     """Stores artifacts on Amazon S3."""
 
     def __init__(
-        self, artifact_uri, access_key_id=None, secret_access_key=None, session_token=None
+        self,
+        artifact_uri,
+        access_key_id=None,
+        secret_access_key=None,
+        session_token=None,
+        addressing_style="path",
+        use_optimized=False,
     ):
         super().__init__(artifact_uri)
         self._access_key_id = access_key_id
         self._secret_access_key = secret_access_key
         self._session_token = session_token
+        self._addressing_style = addressing_style
+        self.bucket, self.bucket_path = self.parse_s3_compliant_uri(self.artifact_uri)
+        self._use_optimized = use_optimized
+        self._region_name = self._get_region_name()
```

Comment:
Instead of interleaving `optimized` and unoptimized code in the same class / file, can we leave `s3_artifact_repo` untouched and put the optimized impl in a separate file? Then, when constructing the artifact repository from URI in `artifact_repository_registry`, we can decide which repo instance to create: `S3ArtifactRepository` or `OptimizedS3ArtifactRepository`.

Reply:
@jerrylian-db Can we also revert all changes to `test_s3_artifact_repo` and introduce a separate `test_optimized_s3_artifact_repo`?

Reply:
I'm not in favor of this because it would mean the duplication of code for non `log_artifact` and `_download_file` methods. If we make an `Optimized` subclass of the `S3ArtifactRepository`, then the `S3ArtifactRepository` itself would still need to be a subclass of `CloudArtifactRepository`. But then, to preserve the old behavior `S3ArtifactRepository` would need to override the `CloudArtifactRepository` superclass methods. In other words, unless we fully duplicate a lot of code, I don't see a way we can make  separate the optimization into a separate file without touching the `S3ArtifactRepository`. I think that the current approach is a lot more simpler. cc: @smurching

Reply:
I would so much rather we duplicate the code than have two largely different implementations interleaved with a bunch of if / else logic :)